src/main/java/io/divolte/server/config/DivolteConfiguration.java

```diff
@@ -1,37 +1,131 @@
 package io.divolte.server.config;
 
 import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.*;
+import io.divolte.server.config.constraint.MappingSourceSinkReferencesMustExist;
+import io.divolte.server.config.constraint.OneSchemaPerSink;
+import io.divolte.server.config.constraint.SourceAndSinkNamesCannotCollide;
 
 import javax.annotation.ParametersAreNonnullByDefault;
 import javax.validation.Valid;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 @ParametersAreNonnullByDefault
+@MappingSourceSinkReferencesMustExist
+@SourceAndSinkNamesCannotCollide
+@OneSchemaPerSink
 public final class DivolteConfiguration {
-    @Valid public final ServerConfiguration server;
-    @Valid public final TrackingConfiguration tracking;
-    @Valid public final JavascriptConfiguration javascript;
-    @Valid public final IncomingRequestProcessorConfiguration incomingRequestProcessor;
-    @Valid public final KafkaFlusherConfiguration kafkaFlusher;
-    @Valid public final HdfsFlusherConfiguration hdfsFlusher;
+    @Valid public final GlobalConfiguration global;
+
+    // Mappings, sources and sinks are all keyed by their name.
+    @Valid public final ImmutableMap<String,MappingConfiguration> mappings;
+    @Valid public final ImmutableMap<String,SourceConfiguration> sources;
+    @Valid public final ImmutableMap<String,SinkConfiguration> sinks;
 
     @JsonCreator
-    private DivolteConfiguration(
-            final ServerConfiguration server,
-            final TrackingConfiguration tracking,
-            final JavascriptConfiguration javascript,
-            final IncomingRequestProcessorConfiguration incomingRequestProcessor,
-            final KafkaFlusherConfiguration kafkaFlusher,
-            final HdfsFlusherConfiguration hdfsFlusher) {
-        this.server = server;
-        this.tracking = tracking;
-        this.javascript = javascript;
-        this.incomingRequestProcessor = incomingRequestProcessor;
-        this.kafkaFlusher = kafkaFlusher;
-        this.hdfsFlusher = hdfsFlusher;
+    DivolteConfiguration(final GlobalConfiguration global,
+                         final Optional<ImmutableMap<String, SourceConfiguration>> sources,
+                         final Optional<ImmutableMap<String, SinkConfiguration>> sinks,
+                         final Optional<ImmutableMap<String,MappingConfiguration>> mappings) {
+        this.global = Objects.requireNonNull(global);
+        this.sources = sources.orElseGet(DivolteConfiguration::defaultSourceConfigurations);
+        this.sinks = sinks.orElseGet(DivolteConfiguration::defaultSinkConfigurations);
+        this.mappings = mappings.orElseGet(() -> defaultMappingConfigurations(this.sources.keySet(), this.sinks.keySet()));
+    }
+
+    public BrowserSourceConfiguration getBrowserSourceConfiguration(final String sourceName) {
```

Comment:
I think the latter is cleaner: `getSourceConfiguration(…)` with the expected class as a parameter.

The purpose of these methods, and not using the map directly, is to enforce the nullability constraints. Basically it's a programmer error, not a user error, to request a source or sink that doesn't exist or has the wrong type. So the preconditions simply enforce this, and the static analysis tools can tell that the result is never `null`. Going directly to the map with `.get(…)` means null-checking the results.